quadratic_simple_constraint.py

```diff
@@ -1,49 +1,63 @@
 """
 Optuna example that optimizes a simple quadratic function with constraints.
 
-In this example, we optimize a simple quadratic function with constraints.
-
-Please check https://optuna.readthedocs.io/en/stable/faq.html#id16 as well.
+This setup is called `constrained optimization`.
+Constrained optimization is useful when we would like to optimize some metrics under
+some constraints. For example, we would like to maximize the accuracy of a deep neural networks
+while guaranteeing that deep neural networks fit on your hardware, e.g. 8GB of memory consumption.
+In this case, constrained optimization aims to yield an optimal solution that satisfied
+such constraints.
+
+Note that Optuna cannot optimize an objective that will not return any results when some
+constraints violate. For example, when we run a memory-intensive algorithm and user sets
+the memory constraint very close to the limit, we may not get any results if the memory constraint
+is violated. However, Optuna cannot handle such situations.
+Please also check https://optuna.readthedocs.io/en/stable/faq.html#id16 as well.
 
 """
 
 import optuna
 
 
-# Define a simple 2-dimensional objective function with constraints.
+# Define a simple 2-dimensional objective function to minimize.
 def objective(trial):
     x = trial.suggest_float("x", -100, 100)
     y = trial.suggest_categorical("y", [-1, 0, 1])
     return x**2 + y
 
 
 # Define a function that returns constraints.
+# The constraints are to satisfy `c1 <= 0` and `c2 <= 0` simultaneously.
+# If we would like to make the constraint like `c1 <= a`,
+# we simply need to return `c1 - a` instead of `c1`.
 def constraints(trial):
     params = trial.params
     x, y = params["x"], params["y"]
-    # c1 <= 0 and c2 <= 0 must be satisfied.
     c1 = 3 * x * y + 10
     c2 = x * y + 30
-    return [c1, c2]
+    return c1, c2
 
 
 if __name__ == "__main__":
-    # We minimize obj1 and maximize obj2.
+    # Let us minimize the objective function with soft constraints above.
     sampler = optuna.samplers.TPESampler(constraints_func=constraints)
     study = optuna.create_study(sampler=sampler)
     study.optimize(objective, n_trials=500, timeout=1)
 
-    best_trial_id, best_value, best_params = None, float("inf"), None
-    for t in study.trials:
-        infeasible = any(c > 0.0 for c in t.system_attrs["constraints"])
-        if infeasible:
-            continue
-        if best_value > t.value:
-            best_value = t.value
-            best_params = t.params.copy()
-            best_trial_id = t._trial_id
-
-    if best_trial_id is None:
-        print("All trials violated the constraints.")
+    print("Number of finished trials: ", len(study.trials))
+
+    feasible_trial_ids = [
```

Comment:
It might be better to avoid using the word `id` since `trial.trial_id` and `trial.number` are different. How about using other words such as `numbers` or `idx` instead.